### DISCUSSION TITLE
Issue with fetching data from JSONB type using array_agg

### DISCUSSION THREAD

**Discussion by balukrishnans** [sqlalchemy/sqlalchemy#7080] [Usage Questions] [answered]
Got one issue while fetching data from a JSONB column using array aggregation. It won't occur in all cases.

``` py
Foo = Table(
    "Foo",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("bar", JSONB),
)

metadata.create_all(engine)

conn = engine.connect()
with conn.begin() as tran:
    i_query = Foo.insert().values(
        [
            {"id": 1, "bar": [{"buz": 1}]}, 
            {"id": 2,"bar": None},
        ]
    )
    conn.execute(i_query)
    s_query = select([func.array_agg(Foo.c.bar).label("agg_val")])
    result=conn.execute(s_query).fetchall()
print(result[0].agg_val)
```

The above code will throuw an error
```
Traceback (most recent call last):
  File "test.py", line 21, in <module>
    print(result[0].agg_val)
  File "/site-packages/sqlalchemy/dialects/postgresql/array.py", line 359, in process
    return self._proc_array(
  File "/site-packages/sqlalchemy/dialects/postgresql/array.py", line 306, in _proc_array
    return collection(
  File "/site-packages/sqlalchemy/dialects/postgresql/array.py", line 307, in <genexpr>
    self._proc_array(
  File "/site-packages/sqlalchemy/dialects/postgresql/array.py", line 289, in _proc_array
    arr = list(arr)
TypeError: 'NoneType' object is not iterable
```
The expected output is `[[{"buz": 1}], None]`. This issue is because of the check in this [line of _proc_array](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/dialects/postgresql/array.py#L318) function.

From the code understood that issue will be there if
1. first element of array_agg result is array/tuple and any of the remaining object is not iterable

And issue won't be there if
1. First item is not array/tuple. In the above example the issue won't be there if change the position of insert values. ie;
``` py
Foo.insert().values(
    [
        {"id": 1,"bar": None},
        {"id": 2, "bar": [{"buz": 1}]}, 
    ]
)
```

Cc: @CaselIT

**Comment by CaselIT** [comment, marked as answer]
Hi,

Thanks for reporting (btw this was better as an issue, but it's fine here also)
I think think what's failing it the heuristic to guess the dimensions of the array. Maybe we could specify the dimension here:
https://github.com/sqlalchemy/sqlalchemy/blob/7023c07aa6d31b3527cb1171410dcb6e8dff3ec3/lib/sqlalchemy/sql/functions.py#L1331-L1348

If you pass the array dimension in the array_agg function does it work? `func.array_agg(Foo.c.bar, tyoe_=Array(Jsonb, dimensions=1))`

thoughts @zzzeek

**Reply by balukrishnans** [reply]
@CaselIT Thanks for your reply and it's working.

**Reply by CaselIT** [reply]
Thanks for testing that. I'll file as an issue, since we should do that automatically (or at least not fail in the array processing)
